x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_control_panels_url_state.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as rt from 'io-ts';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { fold } from 'fp-ts/lib/Either';
+import { constant, identity } from 'fp-ts/lib/function';
+import { ControlGroupInput } from '@kbn/controls-plugin/common';
+import { useUrlState } from '../../../../utils/use_url_state';
+
+export const getDefaultPanels = (dataViewId: string): ControlGroupInput['panels'] =>
+  ({
+    osPanel: {
+      order: 0,
+      width: 'medium',
+      grow: false,
+      type: 'optionsListControl',
+      explicitInput: {
+        id: 'osPanel',
+        dataViewId,
+        fieldName: 'host.os.name',
+        title: 'Operating System',
+      },
+    },
+    cloudProviderPanel: {
+      order: 1,
+      width: 'medium',
+      grow: false,
+      type: 'optionsListControl',
+      explicitInput: {
+        id: 'cloudProviderPanel',
+        dataViewId,
+        fieldName: 'cloud.provider',
+        title: 'Cloud Provider',
+      },
+    },
+  } as unknown as ControlGroupInput['panels']);
+const HOST_FILTERS_URL_STATE_KEY = 'controlPanels';
+
+export const useControlPanels = (dataViewId: string) => {
+  const [urlState, setUrlState] = useUrlState<ControlPanels>({
+    defaultState: getDefaultPanels(dataViewId),
+    decodeUrlState,
+    encodeUrlState,
+    urlStateKey: HOST_FILTERS_URL_STATE_KEY,
+  });
+
+  return {
+    controlPanel: urlState,
+    setControlPanels: setUrlState,
+  };
+};
```

Comment:
**question**: just  a detail, for stateful hooks like `useState` or `useReducer` the React convention is to return the result as a `Array<Value, UpdaterHandler>`. Following this convention, you could basically reduce this custom hook to a configurator custom hook in the form of
```ts
export const useControlPanels = (dataViewId: string) => {
  return useUrlState<ControlPanels>({
    defaultState: getDefaultPanels(dataViewId),
    decodeUrlState,
    encodeUrlState,
    urlStateKey: HOST_FILTERS_URL_STATE_KEY,
  });
};
```
and then use it as 
```ts
const [ controlPanel, setControlPanels ] = useControlPanels(dataViewId);
```
Non-blocking at all, just curious if there is any reason for following this approach :)

Reply:
Good point, I changed this 👍